flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1264,6 +1265,42 @@ public void testCheckpointDeclinedOnClosedOperator() throws Throwable {
         assertEquals(1, harness.getCheckpointResponder().getDeclineReports().size());
     }
 
+    @Test
+    public void testAbortPreviousCheckpointBeforeCompleteTerminateSavepoint() throws Throwable {
+        // given: Marker that the task is finished.
+        AtomicBoolean finishTask = new AtomicBoolean();
+        StreamTaskMailboxTestHarnessBuilder<Integer> builder =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                (env) ->
+                                        new OneInputStreamTask<Integer, Integer>(env) {
+                                            @Override
+                                            protected void finishTask() throws Exception {
+                                                super.finishTask();
+                                                finishTask.set(true);
+                                            }
+                                        },
+                                BasicTypeInfo.INT_TYPE_INFO)
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO);
+        StreamTaskMailboxTestHarness<Integer> harness =
+                builder.setupOutputForSingletonOperatorChain(
+                                new TestBoundedOneInputStreamOperator())
+                        .build();
+
+        // when: Receiving the abort notification of the previous checkpoint before the complete
+        // notification of the savepoint terminate.
+        harness.streamTask.notifyCheckpointAbortAsync(1);
+        harness.streamTask.notifyCheckpointCompleteAsync(2);
```

Comment:
How can this happen? You cannot complete a checkpoint that has not been triggered yet. The order of `notifyCheckpointCompleteAsync` and `triggerCheckpointOnBarrier` is wrong here.

Reply:
I took this order from the test `terminateWithSavepointWithoutComplicationsShouldSucceedAndLeadJobToFinished`. Usually, it indeed happens in normal order, but sometimes I see that the abort happens before the complete.

Reply:
I meant the order of `notifyCheckpointCompleteAsync` and `triggerCheckpointOnBarrier`.

The barrier must arrive first before we can say the checkpoint completed. It's not possible to receive a `notifyCheckpointCompleteAsync` before we've seen a barrier for the checkpoint.

In your testcase you're first sending the `notifyCheckpointCompleteAsync(2)` and after that you're sending a barrier: `triggerCheckpointOnBarrier(...(2)...)`

Reply:
`notifyCheckpointCompleteAsync` and `notifyCheckpointAbortAsync` actually do nothing but add the task to mailbox. When `triggerCheckpointOnBarrier` is called it first of all does its own job and only then it directly execute all waiting mails until the checkpoint is complete. So in fact the `triggerCheckpointOnBarrier` is always first.

Reply:
Still this is based on the internal knowledge of the component. It is still true that those methods can not be called in that order in the production code. I find the current order both confusing and prone to changes in the implementation. I find it also distracting from the real scenario that's being tested here.

I might be wrong, but I truly believe I am not the only one that would be confused by this order.

Reply:
Do I understand correctly, that you are talking about order in the test? Do you mean that we should explicitly call `triggerCheckpointOnBarrier` first? If so please, check the my last commit I a little rewrote the test

Reply:
Yes, I am talking about the order in the test.

Why wasn't it necessary in the previous version to call the `triggerCheckpointOnBarrier` from within the executor? 

Do we need to call the `notifyCheckpointCompleteAsync` and notifyCheckpointAbortAsync` from the executor? I think those methods are called via RPC in the production code and are not executed from the executor thread...

Reply:
Ok, I checked it myself. If we execute it from the test thread it gets stuck in the synchronous savepoint loop. 

I think executing `triggerCheckpointOnBarrier` from the `MailboxExecutor` is fine here. I would not do it for the `*Async` methods.

Reply:
ok, I have removed the mailbox Executor for async calls. Initially, I added it that only for keeping the one style but it indeed didn't have any real function.